### PR TITLE
[feature] default_cli_command for config what command bundler runs when no specific command is provided

### PR DESCRIPTION
**Pull request by jonbarlo**
## What was the end-user or developer problem that led to this PR?

[Closes#8786](https://github.com/rubygems/rubygems/issues/8786)

## What is your fix for the problem, implemented in this PR?

The `default_cli_command` feature allows users to configure what command Bundler runs when no specific command is provided. This feature was introduced in Bundler 4 to change the default behavior from running `bundle install` to displaying usage help. In Bundler 4, the default behavior when running just `bundle` changed from running `bundle install` to displaying usage help. This feature allows users to configure this default behavior.

### Setting the Default Command

Users can configure the default CLI command using Bundler's settings system:

```bash
# Set to run 'install' by default (Bundler 3 behavior)
bundle config set default_cli_command install

# Set to run 'cli_help' by default (Bundler 4 behavior)
bundle config set default_cli_command cli_help
```

### Valid Values

The `default_cli_command` setting accepts only two valid values:

- `install` - Runs `bundle install` when no command is specified (Bundler 3 behavior)
- `cli_help` - Displays usage help when no command is specified (Bundler 4 behavior)

## Make sure the following tasks are checked

- [x] Describe the problem / feature
- [x] Write [tests](https://github.com/rubygems/rubygems/blob/master/doc/bundler/development/PULL_REQUESTS.md#tests) for features and bug fixes
- [x] Write code to solve the problem
- [x] Make sure you follow the [current code style](https://github.com/rubygems/rubygems/blob/master/doc/bundler/development/PULL_REQUESTS.md#code-formatting) and [write meaningful commit messages without tags](https://github.com/rubygems/rubygems/blob/master/doc/bundler/development/PULL_REQUESTS.md#commit-messages)

### CODE REVIEW

File: bundler/.bundle/config [hunk 1]
```diff
@@ -0,0 +1,2 @@
+---
+BUNDLE_DEFAULT_CLI_COMMAND: "cli_help"
```
Comment (deivid-rodriguez) [inline bundler/.bundle/config]: Not sure what this is about, maybe the spec is leaving side effects? We should make sure to clean up after ourselves.

File: bundler/lib/bundler/cli.rb [hunk 2]
```diff
@@ -86,26 +86,8 @@ def initialize(*args)
...
-
-      shell.say "  Primary commands:\n"
-      shell.print_table(primary_commands, indent: 4, truncate: true)
-      shell.say
-      shell.say "  Utilities:\n"
-      shell.print_table(utilities, indent: 4, truncate: true)
-      shell.say
-      self.class.send(:class_options_help, shell)
+      # Use the enhanced man-based help system for consistency
+      help(nil)
```
Comment (deivid-rodriguez) [inline bundler/lib/bundler/cli.rb]: Indeed I mentioned this improvement in https://github.com/rubygems/rubygems/issues/8786, however, I then realized this is intentional, see https://github.com/rubygems/bundler/pull/5725. So let's remove this part.

File: bundler/lib/bundler/cli.rb [hunk 3]
```diff
@@ -485,7 +467,11 @@ def console(group = nil)
     def version
       cli_help = current_command.name == "cli_help"
       if cli_help || ARGV.include?("version")
-        build_info = " (#{BuildMetadata.timestamp} commit #{BuildMetadata.git_commit_sha})"
+        # Only show build info if it's not the default timestamp
+        timestamp = BuildMetadata.timestamp
+        unless timestamp == "1980-01-02"
+          build_info = " (#{timestamp} commit #{BuildMetadata.git_commit_sha})"
+        end
```
Comment (deivid-rodriguez) [inline bundler/lib/bundler/cli.rb]: This should've already been fixed by https://github.com/rubygems/rubygems/pull/8806, are you still seeing the problem?

File: bundler/lib/bundler/settings.rb [hunk 4]
```diff
@@ -416,13 +417,15 @@ def set_key(raw_key, value, hash, file)
 
       key = key_for(raw_key)
 
+      # Always validate, even if the value is already set
+      Validator.validate!(raw_key, converted_value(value, raw_key), hash)
+
```
Comment (deivid-rodriguez) [inline bundler/lib/bundler/settings.rb]: Is this "order change" necessary for this, or a separate improvement?

File: bundler/spec/bundler/settings_spec.rb [hunk 5]
```diff
@@ -351,4 +351,30 @@
       expect(settings["mirror.https://rubygems.org/"]).to eq("http://example-mirror.rubygems.org")
     end
   end
+
+  describe "default_cli_command validation" do
+    let(:settings) { described_class.new }
```
Comment (deivid-rodriguez) [inline bundler/spec/bundler/settings_spec.rb]: I think this may be the issue making the specs leak. You're overriding the global `subject(:settings)` which correctly uses `bundled_app` as the settings root.

File: bundler/.bundle/config (hunk 1 above)
Comment (jonbarlo) [reply bundler/.bundle/config]: Indeed inadvertently created this file as a side effect of running the test with the wrong settings instance, i did not noticed i actually pushed the files either, already removed.

File: bundler/lib/bundler/cli.rb (hunk 2 above)
Comment (jonbarlo) [reply bundler/lib/bundler/cli.rb]: got ya! ok cool i have reverted this change

File: bundler/lib/bundler/cli.rb (hunk 3 above)
Comment (jonbarlo) [reply bundler/lib/bundler/cli.rb]: you are right! already removed this change.

File: bundler/lib/bundler/settings.rb (hunk 4 above)
Comment (jonbarlo) [reply bundler/lib/bundler/settings.rb]: i was just looking to ensure validation always occurs, it is important for the new default_cli_command setting, if someone tried to set default_cli_command to an invalid value the validation would be completely skipped. The eager return on line #423 will take place without ever checking if the value is valid or not, does this makes sense?

File: bundler/spec/bundler/settings_spec.rb (hunk 5 above)
Comment (jonbarlo) [reply bundler/spec/bundler/settings_spec.rb]: yes you are absolutely right !!!

File: bundler/.bundle/config (hunk 1 above)
Comment (deivid-rodriguez) [reply bundler/.bundle/config]: There seems to still be issues with side effects? Now `.bundle/config` is getting removed and `bundler/.bundle/config` with a different configuration is getting added 🤔.

File: bundler/lib/bundler/settings.rb (hunk 4 above)
Comment (deivid-rodriguez) [reply bundler/lib/bundler/settings.rb]: I'm not fully sure. It seems to me that if the key is already set to the wrong value, it doesn't matter if we raise an error since the wrong configuration will still be there?

File: bundler/.bundle/config (hunk 1 above)
Comment (jonbarlo) [reply bundler/.bundle/config]: oh yikes... my bad just fixed this.

File: bundler/lib/bundler/settings.rb (hunk 4 above)
Comment (jonbarlo) [reply bundler/lib/bundler/settings.rb]: yeah that's 100% true already-existing invalid values get skipped and remains, If we want to catch existing invalid configurations, we will need a different approach.

Like validating on config load or access, but that would be a separate feature enhancement therefore i am undoing this change

File: bundler/lib/bundler/settings/validator.rb [hunk 6]
```diff
@@ -74,6 +74,15 @@ def self.validate!(key, value, settings)
           fail!(key, value, "`#{other_key}` is current set to #{other_setting.inspect}", "the `#{conflicting.join("`, `")}` groups conflict")
         end
       end
+
+      rule %w[default_cli_command], "default_cli_command must be either 'install' or 'cli_help'" do |key, value, settings|
+        valid_values = %w[install cli_help]
+        if value.nil?
+          fail!(key, value, "must be one of: #{valid_values.join(", ")}")
```
Comment (deivid-rodriguez) [inline bundler/lib/bundler/settings/validator.rb]: I think here `nil` means "unsetting", and unsetting configurations should be valid and mean "use the default". You may configure `default_cli_command` to `install` and then eventually decide to remove the setting and use the default through `bundle config unset default_cli_command`, and that would be fine.

So I'd remove the above two lines and check only the case when user tries to set a specific invalid value.

Comment (jonbarlo) [reply bundler/lib/bundler/settings/validator.rb]: sounds good @deivid-rodriguez it is done

File: bundler/spec/bundler/settings_spec.rb [hunk 7]
```diff
@@ -351,4 +351,28 @@
...
+    end
+
+    it "rejects invalid values" do
+      expect { settings.set_local("default_cli_command", "invalid") }.to raise_error(
+        Bundler::InvalidOption,
+        /Setting `default_cli_command` to "invalid" failed:\n - default_cli_command must be either 'install' or 'cli_help'\n - must be one of: install, cli_help/
+      )
+    end
+
+    it "rejects nil values" do
```
Comment (deivid-rodriguez) [inline bundler/spec/bundler/settings_spec.rb]: Please update this test to match the implementation and name it `it "accepts nil values"`.

Comment (deivid-rodriguez) [reply bundler/spec/bundler/settings_spec.rb]: Also, the code in the rule checking for `!value.nil?` in order to not raise an error it's not even triggered by this test, because we `Settings#set_local` returns early if the previous value of the setting (`nil` in this case), matches the value passed to `#set_local` (`nil`, too).

So I'd add an extra test `it "accepts nil values for unsetting an existing value of the setting"` that first sets `settings.set_local("default_cli_command", "install")` and then checks that `settings.set_local("default_cli_command", nil)` does not raise an error.

**Review by deivid-rodriguez — CHANGES_REQUESTED** [review]
